src/Domain/Amr/Helpers.cpp

```diff
@@ -95,6 +95,56 @@ ElementId<VolumeDim> id_of_parent(const ElementId<VolumeDim>& element_id,
           element_id.grid_index()};
 }
 
+template <size_t VolumeDim>
+std::vector<ElementId<VolumeDim>> ids_of_children(
+    const ElementId<VolumeDim>& element_id,
+    const std::array<amr::domain::Flag, VolumeDim>& flags) {
+  using ::operator<<;
+  ASSERT(alg::count(flags, Flag::Split) > 0,
+         "Element " << element_id << " has no children given flags " << flags);
+  ASSERT(alg::count(flags, Flag::Join) == 0,
+         "Splitting and joining an Element is not supported");
+  const size_t block_id = element_id.block_id();
+  const size_t grid_index = element_id.grid_index();
+  if constexpr (VolumeDim == 1) {
+    return {{block_id,
+             {{element_id.segment_id(0).id_of_child(Side::Lower)}},
+             grid_index},
+            {block_id,
+             {{element_id.segment_id(0).id_of_child(Side::Upper)}},
+             grid_index}};
+  } else {
+    const auto segment_ids = element_id.segment_ids();
+    std::array<std::vector<SegmentId>, VolumeDim> child_segment_ids;
+    for (size_t d = 0; d < VolumeDim; ++d) {
+      const SegmentId& segment_id = gsl::at(segment_ids, d);
+      gsl::at(child_segment_ids, d) =
+          gsl::at(flags, d) == Flag::Split
+              ? std::vector{segment_id.id_of_child(Side::Lower),
+                            segment_id.id_of_child(Side::Upper)}
+              : std::vector{segment_id};
+    }
+
+    std::vector<ElementId<VolumeDim>> result{};
+    for (const auto segment_id_xi : child_segment_ids[0]) {
+      for (const auto segment_id_eta : child_segment_ids[1]) {
```

Comment:
FYI there's a similar implementation in `ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.cpp`. There I have an `_impl` function that takes a SegmentId, and template specializations for 1D, 2D, and 3D ElementIds that do an increasing number of loops. I find that easier to understand, but up to you.

Reply:
originally I did have 3 functions, but I switched to this as I liked it better...